File: sorts/mergesort/merge_sort.dart
/**
 * Copyright © https://github.com/jarry All rights reserved.
 * @author: jarryli@gmail.com
 * @version: 1.0
 */

merge(List list, int left, int mid, int right) {
  List temp = [];
  temp.length = list.length;
  var i = left;
  var j = mid + 1;
  var k = 0;

  // 合并两个已排序数组。基于中间值，将原数组分为左右两个区间，两个区间是已分别排好序的。
  // 当左侧指针小于中间值(左侧还没有遍历完)且右侧指针小于最右侧(右侧也没有遍历完)
  while (i <= mid && j <= right) {
    // 拿左右的第一个值进行比较，如果左侧小于等于右侧，则将该项移到临时数组中最左侧
    if (list[i] <= list[j]) {
      // 同时左侧指针移动一个位置
      temp[k++] = list[i++];
    } else {
      // 如果右侧的值大于左侧，就把右侧的值添加到临时数组，右侧指针移动一个
      temp[k++] = list[j++];
    }
  }
  print('merge left: $list');
  // 如果左边数组还有剩余数据，就把左侧剩余都放入到原数组后面
  while (i <= mid) {
    temp[k++] = list[i++];
  }
  print('merge right: $list');
  // 如果右侧数组还有剩余数据，把剩下的数据放入到原数组后面
  while (j <= right) {
    temp[k++] = list[j++];
  }

  // 将排序后的元素全部移动到原数组中
  int x = 0;
  while (left <= right) {
    list[left++] = temp[x++];
  }
  print('merge list: $list');
}

mergeSort(List list, int left, int right) {
  // 得到中间值
  int mid = (left + right) ~/ 2;
  // 如果左侧小于右侧则执行合并排序
  if (left < right) {
    // 递归拆分左侧，直到拆分完成
    mergeSort(list, left, mid);
    print('list left: $list');
    // 递归拆分右侧，直到拆分完成
    mergeSort(list, mid + 1, right);
    print('list right: $list');
    // 再合并左右结果
    merge(list, left, mid, right);
  }
  return list;
}

main() {
  print("merge sort start:");
  List data = [4, 3, 2, 1, 5, -1, 10, 39];
  print(mergeSort(data, 0, data.length - 1));
}

/*
// test
jarry@jarrys-MacBook-Pro mergesort % dart --version
Dart SDK version: 2.14.1 (stable) (Wed Sep 8 13:33:08 2021 +0200) on "macos_x64"
jarry@jarrys-MacBook-Pro mergesort % dart merge_sort.dart 
*/

File: sorts/mergesort/merge_sort.go
/**
 * Copyright © https://github.com/jarry All rights reserved.
 * @author: jarryli@gmail.com
 * @version: 1.0
 */
package main

import "fmt"

/**
 * @desc 新建数组，两两比较移动指针实现已排序数组合并
 */
func merge(one []int, two []int) []int {
	var oneLen = len(one)
	var twoLen = len(two)
	// 新建一个空数组，用来保存排序后的结果
	var result = make([]int, oneLen+twoLen)
	i := 0
	j := 0
	k := 0
	// 遍历两个数组，直到其中一个被遍历完成为止
	for i < oneLen && j < twoLen {
		// 比较两个数组的最小项，将小的项添加到空数组中
		// 小项指针后移一位，继续比较最小项，直到一个数组完成
		if one[i] < two[j] {
			result[k] = one[i]
			k++
			i++
		} else {
			result[k] = two[j]
			k++
			j++
		}
	}

	// 复制剩余的第2个数组项
	for i < oneLen {
		result[k] = one[i]
		k++
		i++
	}

	// 复制剩余的第2个数组项
	for j < twoLen {
		result[k] = two[j]
		k++
		j++
	}
	return result
}

func merge_sort(arr []int) []int {
	// 如果数组长度只有1位则终止递归
	if len(arr) <= 1 {
		return arr
	}
	// 得到数组中间值，一分为二
	mid := (len(arr)) / 2
	// 递归合并左侧部分
	left := merge_sort(arr[:mid])
	// 递归合并右侧部分
	right := merge_sort(arr[mid:])
	// 两两合并已排序数组
	return merge(left, right)
}

func main() {
	var arr1 = []int{7, 11, 9, 10, 12, 13, 8}
	var arr2 = []int{-27, 4, 9, 120, 122, 13, 8, 4, 0, 20}
	fmt.Println("merge_sort arr1:")
	fmt.Println(merge_sort(arr1))

	fmt.Println("merge_sort arr2:")
	fmt.Println(merge_sort(arr2))
}

/*
 jarry@jarrys-MacBook-Pro bubblesort % go version
 go version go1.15 darwin/amd64
jarry@jarrys-MacBook-Pro mergesort % go run merge_sort.go
merge_sort arr1:
[7 8 9 10 11 12 13]
merge_sort arr2:
[-27 0 4 4 8 9 13 20 120 122]
jarry@jarrys-MacBook-Pro mergesort %
*/
